---
title: "RNA Sequencing - Training"
author: "Jack McMurray"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include = F}
knitr::opts_chunk$set(echo = T)
knitr::opts_chunk$set(warning = F, message = F)
```


## Differential expression analysis workflow

This Markdown document covers the entire workflow.

### 1. Setting Up    

#### a. Load libraries
```{r}
# install.packages("tidyverse")
library(tidyverse)
# install.packages("pheatmap")
library(pheatmap)

# install.packages("BiocManager")
# BiocManager::install("DESeq2")
library(DESeq2)
# BiocManager::install("apeglm")
library(apeglm)


```

#### b. Load in the data
```{r, include = T}
## Load in data
data <- read.table("../data/Mov10_full_counts.txt", header = T, row.names = 1) 
meta <- read.table("../meta/Mov10_full_meta.txt", header = T, row.names = 1)
```

If we have a look at these, we'll see that the count data is 23,368 rows long and 8 columns. The rows reflect the genes, and the columns the number of samples. 

```{r}
head(data)
head(meta)
```


#### c. Check that the sample names match in both files
* If these didn't match, you could use the `match()` function to ensure they do.
```{r, include = T}
### Check that sample names match in both files
all(colnames(data) %in% rownames(meta))
all(colnames(data) == rownames(meta))
```

#### d. Create the DESeq2 object
* So we take the count data and the meta data tables as input for this. The DESeq object is like a list, which contains lots of different information. It comes pre-built with slots open for various data
* Here, the design refers to the column or columns that we're interested in. In our case, this is sampletype, which has three levels. This tells DESeq2 that for each gene, we want to evaluate gene expression changes with respect to those different levels.
```{r, include = T}
## Create the DESeq2 object
dds <- DESeqDataSetFromMatrix(countData = data, colData = meta, design = ~ sampletype)
```


### 2. Quality Control 
Firstly, we need to normalise the counts like I went through in the seminar last week. Well, we don't actually need to do this for the quality control workflow, all we need is the Size Factors which will be used in the transformation step so we can visualise the data a bit better. To do this, we first estimate the size factors. We can put this back into the dds object since there is a slot for it.

```{r, include = T}
dds <- estimateSizeFactors(dds)
```

* The transformation step uses regularised logarithm of the counts. This step just helps with visualising the data a bit better. We can do this by using the `rlog()` function. 
* This function can be quite slow if doing it for > 20 samples, so if you are performing it on more than 20 then use the `vst()` function instead which performs similar stuff, but is a lot quicker.
* We also set an argument (blind) to TRUE, which performs the transformation blind of the sample condition. This is important to do in quality control since we don't want any transformations that increase the similarity between the sample conditions.


```{r, include = T}
rld <- rlog(dds, blind = T)
```

* Usually quality control is performed using two different metrics - principal component analysis, and hierarchical clustering.

#### a. PCA
DESeq2 has an inbuilt function for producing PCA plots which utilises another package called ggplot2 under the hood. We need only one argument specified, which is the argument to colour these. In this case, we're interested in sampletype.
```{r, include = T}
plotPCA(rld, intgroup = "sampletype")
```
The PCAplot function uses the top 500 most variable genes as default, plotting the first and second principal components. Here, they account for 98% of all variance in the data, so are pretty good measures. We see that each of the sampletypes are near one another...

We could use this bit to investigate other measures of variance as well. For instance, if we had sex in our meta data table. And this is to just demonstrate... 

```{r, include = T}
dds1 <- dds

colData(dds1)$sex <- rep(c("M", "F"), 4)

rld1 <- rlog(dds1, blind = T)
plotPCA(rld1, intgroup = "sex")
```
* We see no patterns in terms of this affecting the data structure. For instance, if all of these on the right were males and all these were females. Then we might use sex as a covariate when we do differential gene expression later on.


#### b. Hierarchical Clustering
DESeq2 doesn't have an inbuilt function for heatmaps, so we're going to use the `pheatmap()` function from the pheatmap package. This packge requires a matrix as it's input, so we have to retrieve that from the rlog transformed values first
* We can do this using a function from SummarisedExperiment, which loads when you load up DESeq2
* We then find the correlations between samples. This is how well they correlate with one another in terms of all genes. This can be achieved using the `cor()` function in base R
```{r}
rld_mat <- assay(rld)
rld_cor <- cor(rld_mat)
```

* We can then plot the output using the pheatmap function

```{r}
pheatmap(rld_cor)
```
* Here we see pretty high correlations across the board within the sample types. 
* Similar to PCA, we're also seeing that the sample types cluster together

* Both of these QC checks have told us that we're good to go ahead with the differential gene expression analysis.

### 3. Differential Gene Expression - Normalising it all
So this is super simple. The way that we perform the normalisation for differential gene expression analysis in R is by one line of code. Well two if we have just started here.

* The first step is to create the matrix. If the QC highlighted any outliers, we could remove them and remake this.

```{r, include = T}
## Create the DESeq2 object
dds <- DESeqDataSetFromMatrix(countData = data, colData = meta, design = ~ sampletype)
```
* The important thing from here is the design. Say whether the sample was male or female impacted our results, we could add this as a covariate by doing
`DESeqDataSetFromMatrix(countData = data, colData = meta, design = ~ sampletype + sex)`

* The second step is literally to just go:

```{r, include = T}
## Create the DESeq2 object
dds <- DESeq(dds)
```
* This runs all the steps that I outlined in the seminar last week. So I'm not going to cover what each of them do again.

You can also output some of the estimates with functions such as `plotDisEsts()`

```{r}
# Plot dispersion estimates
plotDispEsts(dds)
```

### 4. Differential Gene Expression - Getting the results tables

To get the output of the differential gene expression, we now have to build results tables for the various comparisons that we're interested in. 
For this experiment we'd be interested in:
* Control vs Mov10 overexpression
* Control vs Mov10 knockdown

DESeq2 automatically uses the levels as the input for the comparisons. Luckily, if we look at the levels of our sampletypes, "control" is already the first level and therefore will be used for comparison for the other two levels.

```{r}
levels(dds$sampletype)
```

If we wanted to change this we can use the `relevel()` function

```{r}
# relevel(dds$sampletype, ref = "control")
```

To build a results table we use the `results()` functon. To tell DESeq2 which groups we want to compare, we supply the contrasts we would like using the `contrasts` argument. The alpha argument here, is what P-value we determine as significant. By default, this is 0.1 but we'll change this to 0.05.

```{r}
contrasts_oe <- c("sampletype", "MOV10_overexpression", "control")
res_tableOE_unshrunken <- results(dds, contrast = contrasts_oe, alpha = 0.05)
```

For some reason, DESeq2 no longer does the shrinkage stuff that I spoke about in the seminar last week by default. So we have to do it ourselves. We use the `apeglm` package to get the `lfcShrink()` function. Here we also have to define the coefficient of the comparison we're interested in. These are kept within the `resultsNames()` of the object

```{r}
resultsNames(dds)
res_tableOE <- lfcShrink(dds, coef = "sampletype_MOV10_overexpression_vs_control", res = res_tableOE_unshrunken)
```
For the overexpression versus control comparison, this is coefficient 3 or we can use the name directly.


We can next have a quick look at how the results table looks

```{r}
# res_tableOE %>% data.frame() %>% View()
```

Here we see some columns, which we can look at what they mean by using the `mcols()` function. 

```{r}
# mcols(res_tableOE, use.names = T)
```

An important note is that with the log2 FoldChange column. The order of the comparison indicates in which situation it's enriched. So for the MOV10_overexpression versus control if the value is positive, then the gene is enriched in the first. So 
```{r}
# res_tableOE %>% data.frame() %>% View()
```


By default, p values are adjusted using benjamini hochberg correction, but this can be changed in the results function we used earlier.
```{r}
# mcols(res_tableOE, use.names = T)
```


To just get significant genes out, so we can filter. First, we can change the table into a nicer format to work with. Currently it's a DESeqResults object, but we can turn this into a tibble (which is like a dataframe). We can then define some cutoffs, like the 0.05 cutoff we decided earlier. We can also add in a log fold cut off, to increase the stringency. So as we're on a log scale, 0.58 would be a 1.5 fold change between the comparisons, which is a fair cut off.

```{r}
res_tableOE_tb <- res_tableOE %>%
  data.frame() %>%
  rownames_to_column(var = "gene") %>% 
  as_tibble()

padj.cutoff <- 0.05
lfc.cutoff <- 0.58

sigOE <- res_tableOE_tb %>%
        filter(padj < padj.cutoff & abs(log2FoldChange) > lfc.cutoff
               )
```


When we look at the top of these, and order them by decreasing P value. Uderstandably, MOV10 is top of the list
```{r}
sigOE[order(sigOE$padj), ] %>% head()
```

Then if we look in the environment, we're getting 950 genes which are significant. If we took off the log fold change, we now get over 6 and a half thousand differentially expressed genes.


```{r}
sigOE_noLFC <- res_tableOE_tb %>%
        filter(padj < padj.cutoff)
```


### 5. Visualisation - Volcano
Now we have our significant differentially expressed genes, we can do various different visualisation methods. However, perhaps the most used for differential gene expression are volcano plots. These plots have the log-transformed adjusted P-values on the y axis and the log2 fold change values on the x axis

To generate these, we need a column in our results data indicating whether or not the gene is considered differentially expressed based on adjusted p-values

```{r}
res_tableOE_tb <- res_tableOE_tb %>% 
                  mutate(threshold_OE = padj < 0.05 & abs(log2FoldChange) >= 0.58)
```

We can then create a ggplot graph of those, colouring them by whether they're significant or not.

```{r}
ggplot(res_tableOE_tb) +
        geom_point(aes(x = log2FoldChange, y = -log10(padj), colour = threshold_OE)) +
        ggtitle("Mov10 overexpression") +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        #scale_y_continuous(limits = c(0,50)) +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))  
```


## Define contrasts, extract results table and shrink log2 fold changes
```{r}
contrast_kd <-  c("sampletype", "MOV10_knockdown", "control")

res_tableKD <- results(dds, contrast = contrast_kd, alpha = 0.05)

res_tableKD <- lfcShrink(dds, coef = "sampletype_MOV10_knockdown_vs_control", res=res_tableKD)
```





